docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,160 @@
+- The initial `recommend index set` is `{a, b}`, the `remaining index set` is `{c,d}` 
+- Add recommend index set `{a, b}` to the schema and at the same time mute existed index `c`
+- Compile all queries in the workload and add up each query's cost. Assume the sum is *20* and then assign it variable `min_cost`
+- Randomly exchange *1* element between the recommend set and remaining set. Assume *a* and *c* is chosen respectively, now the recommend set is `{c, b}` and the remaining set is `{a, d}`. 
+- Apply recommend index set `{c, b}` to the workload and get a sum cost, say *18*. 
+- As the cost decreased from *20* to *18*, update the recommend set from `{a, b}` to `{c, b}` and update `min_cost` to *18*. Then try swapping 1 element between `{c, b}` and `{a, d}` and execute the re-evaluate process. 
+
+Note that if the sum cost with `{c, b}` is *22* ( greater than 20 ), the next *swap* process should happen between `{a, b}` and `{c, d}` instead of between `{c, b}` and `{a, d}`. Loop the *Swap and Re-evaluate* process until timeout.
+
+## Rational 
+
+- Index advisor for PostgreSQL: [official page](https://www.enterprisedb.com/edb-docs/d/edb-postgres-advanced-server/user-guides/user-guide/9.6/EDB_Postgres_Advanced_Server_Guide.1.55.html), [discussions on index advise tool](https://www.postgresql.org/message-id/CA%2BwPC0OG5Ww2TT1bVbm8B3jhj0yY3dxhmMeOt-EP1DttoXNJuw%40mail.gmail.com)
+
+- [IBM-DB2 index advisor](https://www.ibm.com/support/knowledgecenter/en/SSEPGG_10.1.0/com.ibm.db2.luw.admin.cmd.doc/doc/r0002452.html)
+
+## Compatibility
+
+Above discussion highlights the fact that as a building block of index advisor, TiDB SQL  optimizer needs to provide following three functionalities for index advisor: 
+
+1. **add hypothetic index**. That is, adding some *virtual indexes* to the schema in memory, which in the view of SQL optimizer, seems to be real indexes thus the optimizer will output execution plan with a vision that these virtual indexes do actually exist.  
+
+2. **alter invisible index**. There are situations where some already existed indexes have little affect on given workload. At this time, we need to consider if these *outdated* indexes should be removed. Instead of deleting these indexes inconsiderately, we can make our decisions after observing the query efficiency with these indexes being muted. To achieve this, we could append a keyword *invisible* to target index's metadata in schema to mute it. 
+
+3. **get plan and its cost**. Best physical plan and its cost output by SQL optimizer are essentials to index advisor to recommend indexes for given workload.  
+
+
+## Implementation
+
+### <span id="compatibility">**Support from TiDB SQL optimizer** </span>
+
+1. **add hypothetic indexes to schema** 
+   - **effect**: query should be compiled with *schema* that contains input hypothetic indexes 
+   - **requirements**: these hypothetic indexes should take effects only within current index advisor session  
```

Comment:
Better to use `constraints` instead of `requirements`?

Reply:
Sure~